### DISCUSSION TITLE
Optimisations ideas for Apple M1

### DISCUSSION THREAD

**Discussion by dougallj** [simdjson/simdjson#1658] [Ideas]
I was messing with the simdjson minify benchmark on the M1 from the [_ARM MacBook vs Intel MacBook: a SIMD benchmark_ blog post](https://lemire.me/blog/2020/12/13/arm-macbook-vs-intel-macbook-a-simd-benchmark/) (December, 2020) and came up with a couple of suggestions to improve performance on the M1. Combined, these hit 7.8GB/s on the minify benchmark, bringing performance in line with the Kaby Lake 2017 MacBook Pro (at least on that benchmark). So far I've just been poking through the code and looking at my [instruction tables](https://dougallj.github.io/applecpu/firestorm-simd.html), rather than profiling or looking at counters, so there's probably room for more tuning. (To reproduce my numbers, see [the full diff on gist.github.com](https://gist.github.com/dougallj/5932365ab1ed0d5cb5995383e77c82b1) - note this applies to the [Code used on Daniel Lemire's blog](https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog) repository, not this one.)

I'm posting these as a discussion, as I'm not sure how they perform on other benchmarks and platforms. Let me know if you'd like these as pull requests, or issues.

## Short-circuit if there are no backslashes (~14%)

Uncommenting this line gives a ~14% speedup, and brings the code closer to the x86 implementation:

https://github.com/simdjson/simdjson/blob/8eed8f5155121f2b82deccc754ecc4a16eb51431/src/arm64/dom_parser_implementation.cpp#L122-L125

Hard to beat that performance to characters-changed ratio :)

I'm not entirely certain why this is so effective. (I'm guessing it might allow the branch predictor to break a latency chain?)

## Avoid popcounts (~4%)

Popcount on ARM64 has much worse performance characteristics than on x86, as it makes a roundtrip from integer to SIMD (four instructions, 15c latency on M1 Firestorm). `count_ones` is used extensively in `compress`:

https://github.com/simdjson/simdjson/blob/2abcc35031152ac4a8d867775b035203a8ae7541/include/simdjson/arm64/simd.h#L442-L447

ARM64 `compress` already uses loads to get the popcount of every second byte, which is much more efficient:

https://github.com/simdjson/simdjson/blob/2abcc35031152ac4a8d867775b035203a8ae7541/include/simdjson/arm64/simd.h#L282

If we return the sum of the popcount of both bytes (`pop1 + BitsSetTable256mul2[mask2]`), we can accumulate the popcount as we go instead, for a ~4% speedup:

```cpp
      int pop = this->chunks[0].compress(uint16_t(mask), output);
      pop += this->chunks[1].compress(uint16_t(mask >> 16), output + 16 - (pop >> 1));
      pop += this->chunks[2].compress(uint16_t(mask >> 32), output + 32 - (pop >> 1));
      this->chunks[3].compress(uint16_t(mask >> 48), output + 48 - (pop >> 1));
```

(Note that, confusingly, `pop` is really the popcount multiplied by two.)

Surprisingly, returning `pop` to replace the call to `count_ones` in `json_minifier::next` didn't seem to help. I'm not sure why exactly.

## Faster move-mask replacement (~1.5%)

This is a multiplication-based technique based on 0b0000000000000's work, I believe ([1](https://twitter.com/0b0000000000000/status/1376568414634840065), [2](https://twitter.com/zeuxcg/status/1376754971312488449)). 

I came up with a 32-bit version with two magic numbers that set the top byte of the result to either exactly `mask` or `mask << 4`, and where the low bits don't carry into the top byte when pairs are added together. As such, I add the values together in pairs, then use a 32-byte wide TBL to pull together the bytes of the result. ​On the M1 Firestorm, this saves one cycle of latency (8c to 7c) and one instruction/uop (coincidentally, also 8 to 7), although it does require a second constant register.

```cpp
     ​const uint32x4_t magic = { 0xf0f1f3f8, 0x0f1f3f80, 0xf0f1f3f8, 0x0f1f3f80 };
     ​const uint8x8_t idx = { 3, 7, 11, 15, 19, 23, 27, 31 };

     ​uint8x16x2_t tbl = {
       ​vpaddq_u32(vmulq_u32(this->chunks[0], magic), vmulq_u32(this->chunks[1], magic)),
       ​vpaddq_u32(vmulq_u32(this->chunks[2], magic), vmulq_u32(this->chunks[3], magic)),
     ​};
     ​return vget_lane_u64(vreinterpret_u64_u8(vqtbl2_u8(tbl, idx)), 0);
```

This improves the benchmark by about 1.5%. However, the multiply and TBL2 may be more expensive on other microarchitectures.

## Independent whitespace matching (~1%)

A comment in `json_character_block::classify` explains that minify can't be optimised as well on arm64 as on x64:

https://github.com/simdjson/simdjson/blob/b08818fab9b9153716534b382cc9a18d14248d55/src/arm64/dom_parser_implementation.cpp#L44-L58

Since all whitespace characters are within 32 ascii values of each other, I replaced this whitespace technique with a subtract and a 32-byte TBL:

```cpp
  const uint8x16_t tbl0 = { 0, 0, 0, 0, 0, 0, 0, 0, 0xFF, 0xFF, 0, 0, 0xFF, 0, 0, 0 };
  const uint8x16_t tbl1 = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xFF };
  const uint8x16x2_t tbl = { tbl0, tbl1 };

  uint64_t whitespace = simd8x64<bool>(
        vqtbl2q_u8(tbl, (in.chunks[0] - 1).value),
        vqtbl2q_u8(tbl, (in.chunks[1] - 1).value),
        vqtbl2q_u8(tbl, (in.chunks[2] - 1).value),
        vqtbl2q_u8(tbl, (in.chunks[3] - 1).value)
  ).to_bitmask();
```

Note: when *not* minifying this change would be *worse* in terms of register pressure and throughput, and it's only a 1% difference, although the lower latency in exchange for one more uop might still be a win on the M1.

**Comment by lemire** [comment]
Would you consider making a PR with the first two optimizations (Short-circuit if there are no backslashes and Avoid popcounts)?

It seems like it involves very few lines of code, so it is very nice. We can then assess it collaboratively. I do have an M1 *and* I have access to other ARM cores. An 18% gain in total would be amazing.

I would suggest looking at the smaller optimizations separately.

**Reply by lemire** [reply]
Note that it is entirely credible that we could find low-hanging fruits in the ARM kernel given how little optimization effort we put into it. For one thing, simdjson and its ARM kernel were published long before the M1.

So I am encouraging you... it is likely that we can boost the performance quite a bit with little effort.

**Reply by dougallj** [reply]
Unfortunately the total gain for those two is only 14% - the numbers above were measured by removing each optimisation from my working copy, rather than adding them to the original.

Surprisingly, the popcount change seems to depend on the branch change, so the numbers end up:
* Original: 6.6GB/s
* Popcount change: 6.6GB/s (~0%)
* Branch change: 7.1GB/s (~8%)
* Popcount + branch change: 7.5GB/s (~14%)
* All: 7.8GB/s (~17%)

So that's a bit confusing, sorry! I'd like to figure out what the bottlenecks are and make better sense of this - as you say, it could lead to interesting things :) Until then, I've opened #1659 with the two more-significant changes.

**Comment by dougallj** [comment]
Running out of time this weekend, so this is a little half-baked (sorry) but playing with a standalone (manually inlined) version, this seems to minify at roughly 8.6GiB/s to 9.2GiB/s without branches on the M1 (~3 bytes per cycle): https://gist.github.com/dougallj/45411f4dd5b4e0a3d8e7489bab9ac526

I'm assuming the main change is the `compress` logic - I doubled the number of stores to reduce the amount of work on the SIMD units, and I use another 2kb for a second thintable, but only one popcount roundtrip (and a lot of `UBFX` instructions).

```cpp
// These two lines have the same latency as "count_ones", but give us every popcount
// we need packed into "prefix_sum".
uint64_t popcounts = vget_lane_u64(vreinterpret_u64_u8(vcnt_u8(vcreate_u8(keep))), 0);
uint64_t prefix_sum = popcounts * 0x0101010101010101;

vst1_u8((uint8_t*)output,                             vqtbl1_u8(c0, vcreate_u8(thintable0[ keep        & 0xFF])));
vst1_u8((uint8_t*)&output[popcounts & 0xFF],          vqtbl1_u8(c0, vcreate_u8(thintable1[(keep >>  8) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >>  8) & 0xFF], vqtbl1_u8(c1, vcreate_u8(thintable0[(keep >> 16) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >> 16) & 0xFF], vqtbl1_u8(c1, vcreate_u8(thintable1[(keep >> 24) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >> 24) & 0xFF], vqtbl1_u8(c2, vcreate_u8(thintable0[(keep >> 32) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >> 32) & 0xFF], vqtbl1_u8(c2, vcreate_u8(thintable1[(keep >> 40) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >> 40) & 0xFF], vqtbl1_u8(c3, vcreate_u8(thintable0[(keep >> 48) & 0xFF])));
vst1_u8((uint8_t*)&output[(prefix_sum >> 48) & 0xFF], vqtbl1_u8(c3, vcreate_u8(thintable1[(keep >> 56) & 0xFF])));
output += prefix_sum >> 56;
```

So, yeah - I'll have to test in situ and profile the individual changes, but just writing up this idea while I remember.

**Comment by dougallj** [comment]
The "faster move-mask replacement" change above does a bit better in other benchmarks, with a ~9.5% speedup on `parse_gsoc` and a ~6.9% speedup on `parse_twitter` using clang. There are similarly significant speedups of 5-13% in many of the `bench_ondemand` tests. (There are also a handful of small regressions of less than 1.5%, but these are inconsistent, so seem likely to be noise.)

Unfortunately it almost certainly hurts performance on other ARM64 implementations that have less multiplication throughput, and the speedups are smaller under GCC. But if simdjson targeted Apple CPUs specifically (e.g. with `#ifdef __APPLE__`), we could consider it.

**Reply by lemire** [reply]
>  But if simdjson targeted Apple CPUs specifically (e.g. with #ifdef __APPLE__), we could consider it.

The simdjson library is already built with multiple kernels, so there is no objection in principle to having a kernel that is built specifically for apple platforms. Of course, there is some work involved in doing that (less than one might think, but it takes more than 5 minutes).
